File: leetcode/problems/language-specific/go/543.diameter-of-binary-tree.go
/*
 * @lc app=leetcode id=543 lang=golang
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	diameter := 0
	diameterOfBinaryTreeHelper(root, &diameter)
	return diameter
}

func diameterOfBinaryTreeHelper(root *TreeNode, diameter *int) int {
	if root == nil {
		return 0
	}

	left := diameterOfBinaryTreeHelper(root.Left, diameter)
	right := diameterOfBinaryTreeHelper(root.Right, diameter)
	*diameter = findMax(*diameter, left+right)
	return 1 + findMax(left, right)
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/sessions/2025-10-25_2025-??-??/543.diameter-of-binary-tree.py
#
# @lc app=leetcode id=543 lang=python3
#
# [543] Diameter of Binary Tree
#

# @lc code=start

# * Recursive DFS Solution | O(n) Time | O(h) Space
# * n -> The number of nodes in the tree | h -> The height of the tree


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0

        def helper(root):
            nonlocal diameter

            if not root:
                return 0

            left_node_depth = helper(root.left)
            right_node_depth = helper(root.right)
            diameter = max(diameter, left_node_depth + right_node_depth)
            return 1 + max(left_node_depth, right_node_depth)

        helper(root)
        return diameter


# @lc code=end
